BootloaderCommonPkg/Library/ConsoleOutLib/ConsoleOutLib.c

```diff
@@ -0,0 +1,174 @@
+  @retval >0               The number of bytes written to the frame buffer.
+                           If this value is less than NumberOfBytes, then the write operation failed.
+**/
+STATIC
+UINTN
+EFIAPI
+ConsoleWriteFrameBuffer (
+  IN UINT8     *Buffer,
+  IN UINTN      NumberOfBytes
+  )
+{
+  FRAME_BUFFER_CONSOLE *Console;
+
+  ASSERT (Buffer != NULL);
+
+  Console = (FRAME_BUFFER_CONSOLE *)(UINTN)PcdGet64 (PcdFrameBufferConsoleBase);
+  if (Console == NULL || NumberOfBytes == 0) {
+    return 0;
+  }
+
+  FrameBufferConsoleWrite (Console, Buffer, NumberOfBytes,
+                           Colors[WHITE], Colors[BLACK]);
+
+  return NumberOfBytes;
+}
+
+/**
+  Perform a blanking backspace on the frame buffer console.
+
+  @param  Count     Number of backspaces.
+
+  @retval >=0       Number of backspaces performed. If this value is less
+                    than Count, the operation failed.
+**/
+STATIC
+UINTN
+EFIAPI
+ConsoleBackspaceFrameBuffer (
+  IN UINTN Count
+  )
+{
+  FRAME_BUFFER_CONSOLE *Console;
+  UINTN                 i;
+
+  Console = (FRAME_BUFFER_CONSOLE *)(UINTN)PcdGet64 (PcdFrameBufferConsoleBase);
+  if (Console == NULL || Count == 0) {
+    return 0;
+  }
+
+  UINT8 Space = ' ';
+  for (i = 0; i < Count; i++) {
+    FrameBufferConsoleBackspace (Console);
+    FrameBufferConsoleWrite (Console, &Space, 1, Colors[WHITE], Colors[BLACK]);
+    FrameBufferConsoleBackspace (Console);
+  }
+
+  return Count;
+}
+
+/**
+  Perform a blanking backspace on the serial console.
+
+  @param  Count     Number of backspaces.
+
+  @retval >=0       Number of backspaces performed. If this value is less
+                    than Count, the operation failed.
+**/
+STATIC
+UINTN
+EFIAPI
+ConsoleBackspaceSerialPort (
+  IN UINTN Count
+  )
+{
+  CHAR8 Buffer[32];
+  AsciiSPrint (Buffer, sizeof (Buffer), "\x1b[%dD\x1b[K", Count);
+  SerialPortWrite ((UINT8 *)Buffer, AsciiStrLen (Buffer));
+  return Count;
+}
+
+/**
+  Write data from buffer to the console.
+
+  Writes NumberOfBytes data bytes from Buffer to the virtual console.
+
+  If Buffer is NULL, then ASSERT().
+
+  If NumberOfBytes is zero, then return 0.
+
+  @param  Buffer           Pointer to the data buffer to be written.
+  @param  NumberOfBytes    Number of bytes to write to the console.
+
+  @retval >0               The number of bytes written to the console.
+**/
+UINTN
+EFIAPI
+ConsoleWrite (
+  IN UINT8     *Buffer,
+  IN UINTN      NumberOfBytes
+  )
+{
+  ASSERT (Buffer != NULL);
+
+  if (PcdGet32 (PcdConsoleOutDeviceMask) & ConsoleOutSerialPort) {
+    SerialPortWrite ((UINT8 *)Buffer, NumberOfBytes);
+  }
+  if (PcdGet32 (PcdConsoleOutDeviceMask) & ConsoleOutFrameBuffer) {
+    ConsoleWriteFrameBuffer ((UINT8 *)Buffer, NumberOfBytes);
+  }
+
+  return NumberOfBytes;
+}
+
+/**
+  Perform a blanking backspace on the virtual console.
+
+  @param  Count     Number of backspaces.
+
+  @retval >=0       Number of backspaces performed.
+**/
+UINTN
+EFIAPI
+ConsoleBackspace (
```

Comment:
do we need this ConsoleBackSpace () API ?   Maybe better to keep it as internal function.